File: ts/128.longest-consecutive-sequence.ts
/*
 * @lc app=leetcode id=128 lang=typescript
 *
 * [128] Longest Consecutive Sequence
 */

// @lc code=start
function longestConsecutive(nums: number[]): number {
  const set = new Set(nums);
  let longestStreak = 0;

  for (const num of set) {
    if (!set.has(num - 1)) {
      let currentNumber = num;
      let currentStreak = 1;

      while (set.has(currentNumber + 1)) {
        currentNumber++;
        currentStreak++;
      }

      longestStreak = Math.max(longestStreak, currentStreak);
    }
  }

  return longestStreak;
}
// @lc code=end

File: go/128.longest-consecutive-sequence.go
func longestConsecutive(nums []int) int {
    if len(nums) == 0 {
        return 0
    }
    sort.Ints(nums)
    mx, curr := 1, 1
    for i := 1; i < len(nums); i++ {
        if nums[i] != nums[i-1] {
            if nums[i] == nums[i-1]+1 { 
                curr++
            } else {
                mx = max(mx, curr)
                curr = 1 
            }
        }
    }
    return max(mx, curr)
}

func max(a, b int) int {
    if a > b {
        return a
    }
    return b
}